### PR TITLE
Feat/functional api

### PR DESCRIPTION
**Pull request by S1ro1**
## Summary
Add functional api for functions in `liger_kernel.ops`, mapping them to `liger_kernel.transformers.functional` as mentioned in #151 .
Also adds tests assuring the behaviour between the functional and class variant stays the same.

## Testing Done
Tested the behaviour, comparing the functional equivalent to its class counterpart, added tests to appropriate files in `tests/transformers`.

- Hardware Type: NVIDIA RTX A5500
- [x] run `make test` to ensure correctness
- [x] run `make checkstyle` to ensure code style
- [x] run `make test-convergence` to ensure convergence

```
(.venv) xsirov00@sophie3:~/liger-dev/Liger-Kernel$ make test
python -m pytest --disable-warnings test/ --ignore=test/convergence
=========================================================== test session starts ============================================================
platform linux -- Python 3.10.12, pytest-8.3.2, pluggy-1.5.0
rootdir: /home/xsirov00/liger-dev/Liger-Kernel
collected 267 items

test/transformers/test_auto_model.py .                                                                                               [  0%]
test/transformers/test_cross_entropy.py ................................................................................ss           [ 31%]
test/transformers/test_fused_linear_cross_entropy.py ................                                                                [ 37%]
test/transformers/test_geglu.py ................                                                                                     [ 43%]
test/transformers/test_monkey_patch.py .....                                                                                         [ 44%]
test/transformers/test_rms_norm.py ................................................................................................  [ 80%]
test/transformers/test_rope.py ........................                                                                              [ 89%]
test/transformers/test_swiglu.py ........................                                                                            [ 98%]
test/transformers/test_trainer_integration.py .                                                                                      [ 99%]
test/triton/test_triton_monkey_patch.py ..                                                                                           [100%]

================================================ 265 passed, 2 skipped in 60.80s (0:01:00) =================================================

(.venv) xsirov00@sophie3:~/liger-dev/Liger-Kernel$ make test-convergence
HF_DATASETS_OFFLINE=1 python -m pytest --disable-warnings test/convergence
=========================================================== test session starts ============================================================
platform linux -- Python 3.10.12, pytest-8.3.2, pluggy-1.5.0
rootdir: /home/xsirov00/liger-dev/Liger-Kernel
collected 28 items

test/convergence/test_mini_models.py ..............                                                                                  [ 50%]
test/convergence/test_mini_models_no_logits.py ..............                                                                        [100%]

====================================================== 28 passed in 177.49s (0:02:57) ======================================================

(.venv) xsirov00@sophie3:~/liger-dev/Liger-Kernel$ make checkstyle
flake8 .; flake8_status=$?; \
isort .; isort_status=$?; \
black .; black_status=$?; \
if [ $flake8_status -ne 0 ] || [ $isort_status -ne 0 ] || [ $black_status -ne 0 ]; then \
	exit 1; \
fi
Skipped 1 files
All done! ✨ 🍰 ✨
59 files left unchanged.
```

### CODE REVIEW

**Review by lancerts — COMMENTED** [review]
Nice contribution. lgtm. @ByronHsu @yundai424 for second check.

**Review by yundai424 — APPROVED** [review]
LGTM! you might want to update the branch in order to merge

**Comment by S1ro1** [comment]
Should be good to go now 🫡

**Review by ByronHsu — CHANGES_REQUESTED** [review]
IMO we don't need to write separate tests for functional. this will double the testing time. we can just use pytest patch to make sure the input args are correctly passed

**Comment by ByronHsu** [comment]
As the num of kernels, this will become very time consuming

**Comment by lancerts** [comment]
@S1ro1 after discussion with Byron, lets use fewer/smaller size testcases for testing functional API parity so that test time doesn't increase. Thanks.

**Comment by lancerts** [comment]
For example,
```
@pytest.mark.parametrize(
    "bsz, seq_len, size",
    [
        (2, 2048, 11008),
        (2, 2048, 4096),
        # weird shapes
        (9, 41, 4231),
        (6, 42, 2048),
    ],
)
```

to 
```
@pytest.mark.parametrize(
    "bsz, seq_len, size",
    [
        (2, 2, 8),
        # weird shapes
        (9, 7, 41),
    ],
)
```

**Comment by S1ro1** [comment]
Okay, will rewrite it tomorrow since it's pretty late in my timezone as of now.

**Comment by S1ro1** [comment]
@lancerts have implemented the changes, effectively reducing testing time from ~62sec to 50sec on RTX A5500
Also added the newly implemented layer_norm to the api and wrote tests for that too. Should be good to go now.

File: test/transformers/test_swiglu.py [hunk 1]
```diff
@@ -177,3 +175,49 @@ def test_correctness_phi3mlp(
     )
 
     assert torch.allclose(x1.grad, x2.grad, atol=atol, rtol=rtol)
+
+
+@pytest.mark.parametrize(
+    "bsz, seq_len, size",
+    [
+        (2, 2048, 11008),
```
Comment (lancerts) [inline test/transformers/test_swiglu.py:183]: can we make these shape small as well? ty

**Comment by S1ro1** [comment]
Have apparently changed test sizes for a wrong test, thank you for the catch, should be fixed now.

**Review by ByronHsu — APPROVED** [review]

**Pull request merged**